### PR TITLE
[DDW-697] Automate Theme Management

### PR DESCRIPTION
**Pull request by MarcusHurney**
This PR automates the manual maintenance of the existing Daedalus themes through the use of the `createTheme` object and the `themes:update` script. Using this script, all existing themes will be compared to the `createTheme` object, and if any definitions are missing, then themes with missing definitions can be updated with the needed CSS variables and `createTheme` will also programmatically generate the color values for the new CSS vars.

## How to test (for developers)

Since there are no new variables on `createTheme` that need to be written to the existing 3 themes, when you run `yarn themes:update` in the nix-shell you will should see this in the terminal:

Screen Shot 2019-07-16 at 10 03 01 PM

For the purpose of demoing the `themes:update` script, I've added a CSS variable to `createTheme` called `--theme-button-border-color-demo`: `${border}`

Screen Shot 2019-07-16 at 10 04 54 PM

Running `yarn themes:update` again will now show in the terminal that this variable is missing in the 3 themes and will also show the color value `createTheme` generated for each given theme. It will ask if you want to write this new variable to the themes.

Screen Shot 2019-07-16 at 10 05 37 PM

After entering `yes`, the script will write the variable and values to each theme file, and then run `prettier` on the file so it is formatted properly. You can check the output by running `git diff`.

Screen Shot 2019-07-16 at 10 08 00 PM

The demo variable above has not been included in this PR, but is just an example out how this script can be used. It will also be possible to write an entire theme to a file with this script by taking the full output of `createTheme` (with new color params) and writing it to a new theme file alongside our current 3. This will be used in the next PR for `DDW-757` which adds new themes.

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/messages/GGKFXSKC6)
- [ ] Test

---

### Basics

- [x] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [x] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [x] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [x] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [x] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [x] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [x] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [x] There are no *flow* errors or warnings (`yarn flow:test`)
- [x] There are no *lint* errors or warnings (`yarn lint`)
- [x] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [x] There are no missing translations (running `yarn manage:translations` produces no changes)
- [x] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [x] Japanese text changes are proofread and approved (Junko Oda)
- [x] UI changes look good in all themes (Alexander Rukin)
- [x] Storybook works and no stories are broken (`yarn storybook`)
- [x] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [x] Important parts of the code are properly commented and documented
- [x] Code is properly typed with flow
- [x] React components are split-up enough to avoid unnecessary re-renderings
- [x] Any code that only works in main process is neatly separated from components

### Testing
- [x] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [x] All existing acceptance tests are still up-to-date
- [x] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: source/renderer/app/themes/utils/constants.js [hunk 1]
```diff
@@ -30,3 +36,59 @@ export const CREATE_THEME_MOCK_PARAMS = {
   },
   fonts: DEFAULT_FONTS,
 };
+
+export const CREATE_THEME_OBJ = {
+  ...createReactPolymorphTheme(CREATE_THEME_MOCK_PARAMS),
+  ...createDaedalusComponentsTheme(CREATE_THEME_MOCK_PARAMS),
+};
+
```
Comment (tomothespian) [inline source/renderer/app/themes/utils/constants.js]: @MarcusHurney  Related to "automatism" I mentioned, here can be something like themeParams that you can pass  as  `[CREATE_CARDANO_THEME_PARAMS, CREATE_LIGHT_BLUE_THEME_PARAMS... ]`

File: source/renderer/app/themes/utils/findUpdates.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  return console.log(chalk.hex(color)(message));
+};
+
+// Checks for properties/CSS vars on createThemeObj that don't exist on existing themes
+export const findUpdates = (
+  createThemeOutputs: FindUpdatesParams
+): null | PendingThemesUpdates => {
+  const { cardano, darkBlue, lightBlue } = createThemeOutputs;
+  const pendingThemesUpdates = {};
+
```
Comment (tomothespian) [inline source/renderer/app/themes/utils/findUpdates.js]: @MarcusHurney  Also suggestion to automatise defsToAdd regarding to passed params "${prefix}_THEME_CONFIG" as collection instead manual creation for each of them.

File: source/renderer/app/themes/utils/updateThemes.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import { isEmpty } from 'lodash';
+import { updateTheme } from './createTheme';
+import { CARDANO_THEME_CONFIG } from '../daedalus/cardano';
+import { DARK_BLUE_THEME_CONFIG } from '../daedalus/dark-blue';
+import { LIGHT_BLUE_THEME_CONFIG } from '../daedalus/light-blue';
+import type { PendingThemesUpdates } from '../types';
+
+export const updateThemes = (pendingUpdates: PendingThemesUpdates) => {
+  const { cardanoUpdates, darkBlueUpdates, lightBlueUpdates } = pendingUpdates;
+
```
Comment (tomothespian) [inline source/renderer/app/themes/utils/updateThemes.js]: @MarcusHurney same here - > "Automatism"

**Comment by a-rukin** [comment]
image
noticed this strange line while testing

**Comment by nikolaglumac** [comment]
@a-rukin did you test this in Storybook or Daedalus? 
@IuliaDolishniak can you please test this case @a-rukin is reporting?

**Comment by nikolaglumac** [comment]
This one can only be merged after we get an approval from @IuliaDolishniak and @daniloprates.

File: source/renderer/app/themes/scripts/update.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import { createTheme } from '../utils/createTheme';
+import { findUpdates } from '../utils/findUpdates';
+import { runUpdateThemesCLI } from '../utils/updateThemesCLI';
+import {
+  CREATE_CARDANO_THEME_PARAMS,
+  CREATE_DARK_BLUE_THEME_PARAMS,
+  CREATE_LIGHT_BLUE_THEME_PARAMS,
+} from '../utils/constants';
+
+const findUpdatesParams = {
```
[the remark this answers is not in the document]
Comment (MarcusHurney) [reply source/renderer/app/themes/scripts/update.js]: @tomothespian
Yes, I think you are right about that. I will add this next week possibly. Thanks!

File: source/renderer/app/themes/utils/constants.js (hunk 1 above)
Comment (MarcusHurney) [reply source/renderer/app/themes/utils/constants.js]: This `CREATE_THEME_OBJ` is only needed to compare its keys against the existing themes. The actual values of the keys aren’t important. Not sure if this ties into the array of actual themes params you’ve suggested. We can discuss it further though.

File: source/renderer/app/themes/utils/findUpdates.js (hunk 2 above)
Comment (MarcusHurney) [reply source/renderer/app/themes/utils/findUpdates.js]: I agree! 👍

**Comment by MarcusHurney** [comment]
@yakovkaravelov
The fs module is only being used in dev mode here. The electron window itself is also not open while the script is running. Should be fine.

**Pull request merged**